design/design-doc-composition-functions.md

```diff
@@ -175,32 +200,606 @@ metadata:
+
+Some key differences between the alpha `FunctionIO` and the proposed beta
+`RunFunctionRequest`:
+
+* `observed.resources` and `desired.resources` are a map keyed by resource name,
+  not an array of objects with name fields. The previous pattern was a result of
+  `FunctionIO` attempting to be an idiomatic KRM object, which is not a
+  constraint for `RunFunctionRequest`. This is should make it easier to lookup
+  resources by name when developing Functions.
+* Entries in `desired.resources` can no longer return 'derived' connection
+  secrets or readiness checks, similar to an entry in the P&T resources array.
+  Instead a Function is intended to set XR connection details and/or readiness
+  by mutating them directly.
+
+The package manager is responsible for creating a headless Kubernetes Service
+where each Function's Deployment can be reached. The address of the Service will
+be exposed as the `status.endpoint` of the Function resource. The Service must
+be headless in order for Crossplane's gRPC client to load-balance connections
+when there are multiple Function replicas.
+
+Note that the fact this endpoint is powered by a Service is an implementation
+detail; it may be possible for Functions to be reached (and indeed deployed) by
+other means in future (see [Runtime Configuration](#runtime-configuration)).
+
+```yaml
+apiVersion: pkg.crossplane.io/v1beta1
+kind: Function
+metadata:
+  name: go-templates
+spec:
+  package: xpkg.upbound.io/negz/go-templates:v0.1.0
+status:
+  # The gRPC endpoint where Crossplane will send RunFunctionRequests.
+  endpoint: https://go-templates-9sdfn2
+```
+
+gRPC communication between Crossplane and a Function will be secured by mutual
+transport layer security (mTLS). Crossplane has an established pattern for
+setting up mTLS certificate bundles - we do this today to secure communication
+between the API server and our validation webhooks, and between Crossplane and
+External Secret Stores. We might want to [rethink this][cert-per-entity] in
+future, but to begin with we'll stick with the established pattern, i.e.:
+
+* The Helm chart creates an empty Secret for the client bundle.
+* The core init container bootstraps the client bundle.
+* The package manager creates a Secret for each Function's server bundle.
+* The package manager bootstraps each Function's server bundle.
+
+See [#3884] for an example of this pattern.
+
+The frequency with which Crossplane will call each Function is relative to how
+frequently each XR is reconciled. There are two main triggers to reconcile an XR:
+
+1. A watch-triggered reconcile, i.e. because the XR changed.
+2. A poll-triggered reconcile, by default once per 60 seconds.
+
+XRs rely on poll-triggered reconciliation to promptly correct drift of their
+composed resources. The poll interval can be set using the `--poll-interval`
+flag. The XR controller does not know what kinds of resources it will compose
+when started, so it cannot start a watch for them.
```

Comment:
The function invocation should not depend on the composed objects.

We know that the first function in the pipeline only depends on the XR, claim (we make that available too, don't we?) and the composition, and with that the whole pipeline.

What we need is a way to detect drift of the composed objects in order to trigger another run of the pipeline on demand. Technically, we don't even need a cached image for the trigger (we have to detect deletions and updates to the spec of composed objects).

----
With that said, it would be good to clearly call out the assumptions a function author can depend on. Being called every 60s should not be one of them IMO. Determinism for a defined set of inputs should be a requirement for every function. That keeps the freedom for us to improve the implementation later.

Reply:
I learned today that compositions are not well-founded in the sense that they can depend on siblings from the life-system through `ToCompositeFieldPath`. We are actually not really building just compositions, but at the same time they are transitions of a state machine. Mind blowing and frightening.

Our innnocent looking sentence in the docs:
> FromCompositeFieldPath and ToCompositeFieldPath patches can also take a wildcarded field path in the toFieldPath parameter and patch each array element in the toFieldPath with the singular value provided in the fromFieldPath.

Reply:
Couldn't both work?

There's a section on potentially watching composed resources under "Future Considerations". Also https://github.com/crossplane/crossplane/issues/4316.